torch/fx/experimental/symbolic_shapes.py

```diff
@@ -186,79 +176,217 @@ def magic_impl(self, other):
             if isinstance(other, PySymInt):
                 other = other.expr
             # TODO: consider constant prop here
-            return PySymInt(func(self.expr, other), self.shape_env)
+            expr = self.shape_env.replace(self.expr)
+            other = self.shape_env.replace(other)
+            out = func(expr, other)
+            out = sympy.expand(out)
+            return PySymInt(out, self.shape_env)
         return magic_impl
 
+    _func = lru_cache(256)(_func)
     # this should be wrapped transparently into torch.SymIntNode
     setattr(PySymInt, method, _create_magic_impl(_func))
     setattr(PySymInt, f"__{method}__", _create_magic_impl(_func))
     if method in reflectable_magic_methods:
         setattr(PySymInt, f"__r{method}__", _create_magic_impl(_func))
 
+def _lru_cache(fn, maxsize=None):
+    """
+    Wrapper around lru_cache that clears when new info about shapes has been
+    updated.
+
+    Use lru_cache if the output is always the same, regardless of the
+    constraints we know now (i.e. evaluate_expr)
+
+    Use _lru_cache otherwise.
+    """
+    fn_cache = lru_cache(maxsize)(fn)
+    prior_key = None
+
+    @functools.wraps(fn)
+    def wrapper(self, *args, **kwargs):
+        nonlocal prior_key
+        if prior_key != self._get_key():
+            prior_key = self._get_key()
+            fn_cache.cache_clear()
+        return fn_cache(self, *args, **kwargs)
+
+    wrapper.cache_info = fn_cache.cache_info  # type: ignore[attr-defined]
+    return wrapper
+
+
+
 class ShapeEnv(object):
     def __init__(self):
         self.guards = []
-        self.shape_env = {}
-
-    def create_symint(self, name, val, shape_env=None):
+        self.var_to_val = {}
+        self.replacements = {}
+        self.divisible = {}
+
+    def _get_key(self):
+        """
+        Defines the current "state" of the guards we've accumulated in this ShapeEnv.
+        Determines when we need to invalidate our cache
+        """
+        return (len(self.replacements), len(self.divisible))
```

Comment:
This implies that replacements/divisible must be monotonically increasing. Doesn't that imply that these sets can get large? When variables resolve to integers, it seems to me that we ought to be able to drop them from the sets entirely (after eliminating them). This is where a mutating unification style algorithm would be more convenient

Reply:
Actually, this is a good point.

We do reduce the size of these, so this logic is wrong.

Reply:
Oh well, I'll fix it in another PR. Having optimal simplification isn't actually strictly necessary I think - it's more of a performance optimization.

Since we can always simplify our guards after tracing.